pypeit/images/rawimage.py

```diff
@@ -17,401 +21,746 @@
+               Generate a cosmic-ray mask
 
         Args:
-            par (:class:`pypeit.par.pypeitpar.ProcessImagesPar`):
+            par (:class:`~pypeit.par.pypeitpar.ProcessImagesPar`):
                 Parameters that dictate the processing of the images.  See
                 :class:`pypeit.par.pypeitpar.ProcessImagesPar` for the
                 defaults.
             bpm (`numpy.ndarray`_, optional):
-            flatimages (:class:`pypeit.flatfield.FlatImages`):
-            bias (`numpy.ndarray`_, optional):
-                Bias image
-            slits (:class:`pypeit.slittrace.SlitTraceSet`, optional):
-                Used to calculate spatial flexure between the image and the slits
+                The bad-pixel mask.  This is used to *overwrite* the default
+                bad-pixel mask for this spectrograph.  The shape must match a
+                trimmed and oriented processed image.
+            flatimages (:class:`~pypeit.flatfield.FlatImages`, optional):
+                Flat-field images used to apply flat-field corrections.
+            bias (:class:`~pypeit.images.pypeitimage.PypeItImage`, optional):
+                Bias image for bias subtraction.
+            slits (:class:`~pypeit.slittrace.SlitTraceSet`, optional):
+                Used to calculate spatial flexure between the image and the
+                slits, if requested via the ``spat_flexure_correct`` parameter
+                in :attr:`par`; see
+                :func:`~pypeit.core.flexure.spat_flexure_shift`.  Also used to
+                construct the slit illumination profile, if requested via the
+                ``use_illumflat`` parameter in :attr:`par`; see
+                :func:`~pypeit.flatfield.FlatImages.fit2illumflat`.
+            debug (:obj:`bool`, optional):
+                Run in debug mode.
+            dark (:class:`~pypeit.images.pypeitimage.PypeItImage`):
+                Dark image
 
         Returns:
-            :class:`pypeit.images.pypeitimage.PypeItImage`:
-
+            :class:`~pypeit.images.pypeitimage.PypeItImage`: The processed
+            image.
         """
+        # TODO: Reset steps to all be false at the beginning of the function?
+        # If we're careful with book-keeping of the attributes (i.e., resetting
+        # image to rawimage, etc), calling process multiple times could be a way
+        # for developers to *re*process an image multiple times from scratch
+        # with different parameters to test changes.
+
+        # Set input to attributes
         self.par = par
-        self._bpm = bpm
-        
-        # Get started
-        # Standard order
-        #   -- May need to allow for other order some day..
-        if par['use_pattern']:  # Note, this step *must* be done before use_overscan
+        if bpm is not None:
+            self._bpm = bpm
+
+        # Check the input
+        # TODO: This may be too strict for how the code is used.  I.e., given
```

Comment:
I prefer we keep as errors until we find it too onerous.

Reply:
Agreed.  Comment removed.